Adds a new DAG to test "Jobset Ready Healthiness" metric

QuinnMMcGarry opened the pull request:
# Description
This change adds a new DAG that automatically checks if the correct number of replicas appear for a jobset's "Ready" status.

# Tests
- DAG Name: jobset_healthiness_ready
- Airflow test runs: https://b371d513bb74418681c3aab7b27bbf12-dot-us-east1.composer.googleusercontent.com/dags/jobset_healthiness_ready/grid?tab=logs&dag_run_id=manual__2025-12-23T20%3A36%3A17.106112%2B00%3A00&task_id=v6e.cleanup_node_pool__1

## Airflow / Composer Environment
- GCP Project: cienet-cmcs
- Composer Environment: tpu-obs-testing
- Composer Version: 2.13.1

## Required Variables
- Cluster Information (This DAG requires an existing cluster)

  - `PROJECT_ID`: cienet-cmcs
  - `CLUSTER_NAME`: tpu-observability-automation-prod
  - `LOCATION`: us-central1
  - `NODE_LOCATIONS`: us-central1-b

- Node Pool / Machine Configuration

  - `NUM_NODE_POOL`: 2
  - `NODE_POOL_NAMES`: jobset-healthiness-ready-v6e / jobset-healthiness-ready-v6e-2
  - `NUM_NODES`: 4
  - `MACHINE_TYPE`: ct6e-standard-4t
  - `TPU_TOPOLOGY`: 4x4


# Checklist
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run one-shot tests and provided workload links above if applicable. 
- [X] I have made or will make corresponding changes to the doc if needed.

alfredyu-cienet committed:
[cienet-private] Enable workflow checks for the primary working branches

QuinnMMcGarry committed:
Adds a new DAG to test "Jobset Ready Healthiness" metric

- Adds an entire new DAG to test this metric automatically
- Adds the "validate_replica_number" function
- Adds a "READY_TPU" Jobset in jobset_util.py

-Fix automatic cluster naming script

-Created "validate_jobset_replica_number" funtion in jobset_util.py
-Changed workload name to "IDLE_READY_TPU_20M"
-Changed schedule time and added statement to only run on PROD environment

Changed to new nodepool creation format

Remove unused DAG flow

Fix format/linting

Linting fix

Delete dags/dags/tpu_observability directory

Delete dags/airflow_monitoring.py

alfredyu-cienet commented on dags/tpu_observability/jobset_healthiness_ready.py:
```diff
@@ -0,0 +1,225 @@
+with models.DAG(  # pylint: disable=unexpected-keyword-arg
+    dag_id="jobset_healthiness_ready",
+    start_date=datetime.datetime(2025, 8, 10),
+    schedule="00 03 * * *",
```
avoid schedule on non-prod composer

alfredyu-cienet replied:
1. Prevent scheduling on non-PROD Composer environments.
2.	Schedule near—but not overlapping with—our other DAGs.

QuinnMMcGarry commented on dags/tpu_observability/jobset_healthiness_ready.py:
```diff
@@ -0,0 +1,225 @@
+
+  for machine in MachineConfigMap:
+    config = machine.value
+    cluster_info = node_pool.Info(
```
I have addressed the other changes locally. Should I wait for a rebase before pushing?

QuinnMMcGarry commented:
> code base outdated
> 
> clue#1: https://github.com/CIeNET-International/ml-auto-solutions-3/pull/156/commits clue#2: <img alt="screenshot" width="933" height="352" src="https://private-user-images.githubusercontent.com/205298813/530590206-7f4c0f17-d3a7-456d-8154-ae97aa34fb98.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjcwMzY3MjEsIm5iZiI6MTc2NzAzNjQyMSwicGF0aCI6Ii8yMDUyOTg4MTMvNTMwNTkwMjA2LTdmNGMwZjE3LWQzYTctNDU2ZC04MTU0LWFlOTdhYTM0ZmI5OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIyOVQxOTI3MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYjY2YzRjZTFmYzdmOWZiZjgwYTkxZmE0MWVkZGFmNjZhYzU4MzBhNTM4ZmQ5NjQ1ZGNmOGIzODUwMGM2ZjZhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.lXuaPpvs2ewZ-5FoCTFLy7_XtpTxdiiuS0aixyTCmKI">

I used `git rebase --onto origin/tpu-obs/dev 7da7f1803bc5323c0d5fad35d0995c05daf46a3c` (my original PR hash), resolved the conflicts, then saved my changes and force pushed them

alfredyu-cienet commented:
This has happened too many times, and I’ve already outlined multiple ways to double-check. This is critical—you are **_required_** to double-check using the methods already provided (PR "Commits" page, git graph, git CLI). Do not submit another revision until you have verified your branch end-to-end.

By “correct,” I mean:
1. the base codebase/branch is correct
2. **_EACH_** commit is correct

**Must perform these checks BEFORE and AFTER submitting any new revision.**
**Must perform these checks BEFORE and AFTER submitting any new revision.**
**Must perform these checks BEFORE and AFTER submitting any new revision.**
---
```
git fetch origin
git checkout tpu-obs/user/QuinnMMcGarry/jobset-healthiness-ready
git reset --hard origin/tpu-obs/user/QuinnMMcGarry/jobset-healthiness-ready
git log tpu-obs/dev..HEAD --stat
```
<img width="1162" height="1118" alt="截圖 2025-12-30 09 25 41" src="https://github.com/user-attachments/assets/8f2a1e8e-615b-4792-9503-504478b5e08b" />

QuinnMMcGarry commented:
> This has happened too many times, and I’ve already outlined multiple ways to double-check. This is critical—you are **_required_** to double-check using the methods already provided (PR "Commits" page, git graph, git CLI). Do not submit another revision until you have verified your branch end-to-end.
> 
> By “correct,” I mean:
> 
> 1. the base codebase/branch is correct
> 2. **_EACH_** commit is correct
> 
> ## **Must perform these checks BEFORE and AFTER submitting any new revision.**
> **Must perform these checks BEFORE and AFTER submitting any new revision.**
> **Must perform these checks BEFORE and AFTER submitting any new revision.**
> ```
> git fetch origin
> git checkout tpu-obs/user/QuinnMMcGarry/jobset-healthiness-ready
> git reset --hard origin/tpu-obs/user/QuinnMMcGarry/jobset-healthiness-ready
> git log tpu-obs/dev..HEAD --stat
> ```
> 
> <img alt="截圖 2025-12-30 09 25 41" width="1162" height="1118" src="https://private-user-images.githubusercontent.com/205298813/530868282-8f2a1e8e-615b-4792-9503-504478b5e08b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjcxMDgyMTgsIm5iZiI6MTc2NzEwNzkxOCwicGF0aCI6Ii8yMDUyOTg4MTMvNTMwODY4MjgyLThmMmExZThlLTYxNWItNDc5Mi05NTAzLTUwNDQ3OGI1ZTA4Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIzMFQxNTE4MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MDBjYzMzZGIzNjBiYmFjMzRiNjM5ODUxMGZkZGU5Y2MzZTJhOTAwNGI0NmU5NjgzMzFkZmFkMzNhOWE1Yjk4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.5XhDE7zGUATOkC68e_SAkZ84Vw06FPKr4DU1GNRMxDc">

You have told me *when* my commits are incorrect and how I *should* be committing them, but not necessarily *why* they are incorrect. In this case I can see there is a commit with an incorrect author and date, I should have seen that it was out of place and tried to fix that. However, it was after rebasing using the `rebase --onto` method you showed me, so I am not sure how it occurred. 

Also, for example, what about my current commit? It is rebased onto the correct branch, and I can see that the code changes are correct. The commit itself looks fine to me, but it has removed the original workflow commit from the PR itself. Is that okay? I am not sure what caused it to happen.
<img width="3024" height="1544" alt="image" src="https://github.com/user-attachments/assets/2d166dc9-da39-420a-a9bf-4705c6b4e330" />
<img width="2128" height="312" alt="image" src="https://github.com/user-attachments/assets/963e3b98-dbe0-4fa9-bd6f-065fa8bb0ac4" />

For that matter, I am also not extremely familiar with git tools, so to me it has made more sense to be able to submit the actual code to be reviewed as soon as it is ready. During which time I can figure out the git issues I am having. I am not ignoring the code changes made in the /dev branch.

QuinnMMcGarry committed:
Add "IDLE_READY_TPU_20M" workload to jobset_util.py

QuinnMMcGarry commented:
> Before I left about two weeks ago, I repointed the old working branch to Google’s master branch in case a production hotfix was needed.
> 
> Now that I’m back and we have a new working branch, please rebase your branch correctly—verify the base branch and merge target, and ensure the commit history and metadata are clean.
> 
> You may encounter failures depending on which commands you use—there is no one-size-fits-all command. The key is to maintain sufficient traceability so you and others can troubleshoot and fix issues properly.
> 
> Please also note that we review each PR with its associated metadata as a whole to keep the process efficient and avoid back-and-forth or duplicate effort. Authors are expected to submit complete, self-contained PRs; partial reviews shift the author’s work onto reviewers and fragment discussion. End-to-end reviews also preserve a single source of truth—feedback stays in one place rather than splitting into conflicting threads.

I rebased onto origin/dev, and have ensured Git CLI, GitGraph, and GitHub UI, all show a total of 4 commits, in the correct order, with the Description/Author/Hash. And all Git checks pass.

Does the Git committing look fine as it is?

QuinnMMcGarry committed:
Removed "retries" from nodepool creation and deletion

QuinnMMcGarry committed:
Refactor to use new
-selector task
-new global variable syntax
-chain instead of ">>"

QuinnMMcGarry committed:
Added "jobset_rollback_ttr" to REGISTERED_DAGS in scheduling_helper.py

QuinnMMcGarry committed:
Reintroduce the "validate_jobset_replica_number" function in the jobset_util.py file

Refactor for new DAG formatting

QuinnMMcGarry committed:
Change the "validate_jobset_replica_number" function to work using the
new jobset storage method. Since "jobset_confifg.replicas" cannot be
accessed in DAG itself.

alfredyu-cienet commented on dags/tpu_observability/jobset_healthiness_ready.py:
```diff
@@ -0,0 +1,197 @@
+            task_id="node_pool_2",
+        )(
+            node_pool=cluster_info_2,
+        )
```
question, what's the purpose of preparing a second node pool

QuinnMMcGarry replied:
Yes, in this case it will remain idle and not affect the number of replicas. But, since the replica number is configured by the default value in tpu_observability_jobset_config.yaml, which is "1", the process will still succeed.

What needs to be decided is whether to actually add a 2nd replica, and declare that in the config file, or to remove the 2nd node-pool from this version and test using only 1. 

The names can also be changed.

alfredyu-cienet replied:
@QuinnMMcGarry thanks for the clarification 

@chiajunglien please check this thread and let me know your progress in taking this over

QuinnMMcGarry commented on dags/tpu_observability/jobset_healthiness_ready.py:
```diff
@@ -0,0 +1,197 @@
+    GCS_CONFIG_PATH,
+    GCS_JOBSET_CONFIG_PATH,
+)
+from dags.common.scheduling_helper.scheduling_helper import SchedulingHelper, get_dag_timeout
```
I took this line from a DAG already in the [dev branch](https://github.com/CIeNET-International/ml-auto-solutions-3/blob/e1c48aaacfb5973902cd4b69c56ebf8a1ba7d520/dags/tpu_observability/jobset_ttr_rollback.py#L33), and it wasn't changed by either GitHub or VS linters. Should I still change it?

chiajunglien committed:
Refactor

chiajunglien committed:
fix

alfredyu-cienet commented:
covered by #256
